### ISSUE TITLE
Images on screen are occasionally out of sync with playlist

### ISSUE THREAD

**Issue by tonycpsu** [mpv-player/mpv#7989]
### Important Information

Provide following Information:
- mpv version: bb23188
- macOS Version: 10.5.15
- Source of the mpv binary or bundle: Homebrew
- If known which version of mpv introduced the problem: 3b13a47 according to #7247 
- Possible screenshot or video of visual glitches: ![mpv_image_desync](https://user-images.githubusercontent.com/914142/89743497-a3a6d200-da71-11ea-989a-1e974af36bee.gif)

### Reproduction steps

Same test reproducer as #7247, and same results as shown above.  The blue image should have been shown after hitting right arrow three times, but purple is shown, with the filename of the blue image shown in the terminal window.  Similar thing happens again toward end of the clip.

### Expected behavior

Images displayed stay in sync with the items in the playlist

### Actual behavior

Images occasionally get out of sync.  Workaround is to resize window or cycle fullscreen, which then displays the correct image.

### Log file

[output.txt](https://github.com/mpv-player/mpv/files/5048557/output.txt)

### Sample files

Generated by test reproducer.

**Comment by Akemi** [comment]
yeah that's totally not the issue described in the original  issue you linked. what you see is completely intended.

you use a hdpi display, what i can see, with a 2x scale. the script uses --video-unscaled (why the fuck one would use this for normal playback is beyond me), eg it doesn't scale if the window is bigger than  the actual video. for the 400x400 image for example a 800x800 window is requested. the black borders you see is just that difference.


use --hidpi-window-scale=no if you don't want that behaviour.

**Comment by tonycpsu** [comment]
Wait, what?  This isn't about scale at all.  The image showing after hitting right three times should be the blue image (the fourth one in the playlist) but it's the purple one (the third image in the playlist.)  I'm not concerned about the scale of the window at all in this case, and the screencast shows clearly that the playlist items shown in the console output and the playlist items shown in the MPV window are out of sync.

**Comment by tonycpsu** [comment]
Frame-by-frame, since there seems to be something lost in translation here: 

![image](https://user-images.githubusercontent.com/914142/89814260-dc49b880-db10-11ea-8ad3-12e14d25bee9.png)

![image](https://user-images.githubusercontent.com/914142/89814287-e79ce400-db10-11ea-8218-d2e61c826236.png)

![image](https://user-images.githubusercontent.com/914142/89814300-ecfa2e80-db10-11ea-87fc-464bda6487a0.png)

![image](https://user-images.githubusercontent.com/914142/89814315-f2f00f80-db10-11ea-82f2-a46a1cce64f7.png)

**Comment by Akemi** [comment]
ignore my inability to properly read, might be that summer heat and having 32°C in my room. i can't reproduce this at all. i assume this is some kind of race condition where the redraw update comes before the actual content changes. might be a libmpv issue?

lets try something if you can build mpv yourself. try changing this function https://github.com/mpv-player/mpv/blob/master/video/out/cocoa-cb/video_layer.swift#L231-L239 to and give me a screencast fo that, i wonder how those prints sync with the output:

```
    func update(force: Bool = false) {
        print("update1 \(force)")
        if force { forceDraw = true }
        queue.async {
            if self.forceDraw || !self.inLiveResize {
                print("update2")
                self.needsFlip = true
                self.display()
                print("update3")
            }
        }
    }
```

on a side note, i dislike how the filenames has resolutions instead of the colour in the name. makes it a bit harder to analyse.

**Comment by Akemi** [comment]
i guess i found the issue. for testing could you change this https://github.com/mpv-player/mpv/blob/master/video/out/cocoa-cb/window.swift#L299-L306 function to:

```
    func updateFrame(_ rect: NSRect) {
        if rect != frame {
            let cRect = frameRect(forContentRect: rect)
            unfsContentFrame = rect
            setFrame(cRect, display: true)
        }
        cocoaCB.layer?.update(force: true)
    }
```

**Comment by tonycpsu** [comment]
Will this do?  I changed the script to make the filename the playlist index (starting with 1) and the color of the image.  The screenshot again shows the blue file with the purple image displayed, and your debug print statements in the console.

![image](https://user-images.githubusercontent.com/914142/89816603-dbb32100-db14-11ea-9440-c5d3b526bce2.png)

**Comment by tonycpsu** [comment]
Still happening for me with the `updateFrame` change.

**Comment by tonycpsu** [comment]
If I add  a `print` statement to `updateFrame`, I notice it isn't called when the window size changes, which I suppose is expected.  I also noticed that the problem can be reproduced even if a `--geometry` argument is passed, making me think this doesn't have anything to do with the window resizing.

**Comment by Akemi** [comment]
something is very unexpected. i would expected at least one "update1 true" in the output, at the point where the window size changes. i also get 5 calls to the update function per file, you only do get 4. which is most likely the source of the problem.

can you change this line https://github.com/mpv-player/mpv/blob/master/video/out/cocoa_cb_common.swift#L94 to 
```
self.layer?.update(force: true)
```

additionally to the print output change. also a screenshot of the output like above would be nice, even if it fixed the problem.

**Comment by tonycpsu** [comment]
Should I leave the `updateFrame` change in as well?

**Comment by Akemi** [comment]
you can leave that one out.

**Comment by tonycpsu** [comment]
Here are two screencasts, both with `--geometry=800x600` added to `test.sh`.

![mpv_image_desync_debug](https://user-images.githubusercontent.com/914142/89819465-8299bc00-db19-11ea-95ed-be839929b572.gif)
The first one is just with your debug prints, and shows the problem twice -- once early on where the image should be #2 (green) but is #1 (orange), and a second time when it should be #3 (purple) but is #4 (blue):

In this one, I've applied your `force: true` change, and cannot reproduce the problem with that change applied.

![mpv_image_desync_debug_force](https://user-images.githubusercontent.com/914142/89819499-8fb6ab00-db19-11ea-90cf-33f3339daa04.gif)

**Comment by Akemi** [comment]
thanks, i think that's enough information for me to narrow down and maybe even reproduce the problem. though i don't think i can do that soon, earliest maybe at the weekend.
